components/ntp_background_images/browser/ntp_sponsored_rich_media_source.cc

```diff
@@ -0,0 +1,157 @@
+}
+
+void NTPSponsoredRichMediaSource::StartDataRequest(
+    const GURL& url,
+    const content::WebContents::Getter& wc_getter,
+    GotDataCallback callback) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+
+  const std::string request_path = URLDataSource::URLToRequestPath(url);
+  const std::optional<base::FilePath> file_path =
+      MaybeGetFilePathForRequestPath(request_path);
+  if (!file_path) {
+    // Deny access.
+    content::GetUIThreadTaskRunner({})->PostTask(
+        FROM_HERE, base::BindOnce(std::move(callback),
+                                  scoped_refptr<base::RefCountedMemory>()));
+
+    return;
+  }
+
+  // Allow access.
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()},
+      base::BindOnce(&ReadFileToString, *file_path),
+      base::BindOnce(&NTPSponsoredRichMediaSource::ReadFileCallback,
+                     weak_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void NTPSponsoredRichMediaSource::ReadFileCallback(
+    GotDataCallback callback,
+    std::optional<std::string> input) {
+  if (!input) {
+    return std::move(callback).Run(scoped_refptr<base::RefCountedMemory>());
+  }
+
+  std::move(callback).Run(
+      new base::RefCountedBytes(base::as_byte_span(*input)));
+}
+
+std::string NTPSponsoredRichMediaSource::GetMimeType(const GURL& url) {
+  std::string mime_type;
+  const base::FilePath::StringType file_path_extension =
+      base::FilePath::FromUTF8Unsafe(url.path_piece()).Extension();
+  if (!file_path_extension.empty()) {
+    net::GetWellKnownMimeTypeFromExtension(file_path_extension.substr(1),
+                                           &mime_type);
+  }
+
+  return mime_type;
+}
+
+bool NTPSponsoredRichMediaSource::AllowCaching() {
+  return false;
+}
+
+std::string NTPSponsoredRichMediaSource::GetContentSecurityPolicy(
+    network::mojom::CSPDirectiveName directive) {
+  switch (directive) {
+    case network::mojom::CSPDirectiveName::FrameAncestors:
+      return std::string("frame-ancestors ") + kBraveUINewTabURL + ";";
+    case network::mojom::CSPDirectiveName::Sandbox:
+      return std::string("sandbox allow-scripts;");
+    case network::mojom::CSPDirectiveName::DefaultSrc:
+      return std::string("default-src 'none';");
+    case network::mojom::CSPDirectiveName::BaseURI:
+      return std::string("base-uri 'none';");
+    case network::mojom::CSPDirectiveName::FormAction:
+      return std::string("form-action 'none';");
+    case network::mojom::CSPDirectiveName::ScriptSrc:
+      return std::string("script-src 'self';");
+    case network::mojom::CSPDirectiveName::StyleSrc:
+      return std::string("style-src 'self';");
+    case network::mojom::CSPDirectiveName::ImgSrc:
+      return std::string("img-src 'self';");
+    case network::mojom::CSPDirectiveName::MediaSrc:
+      return std::string("media-src 'self';");
+    default:
+      return content::URLDataSource::GetContentSecurityPolicy(directive);
```

Comment:
not sure if we want to do this because it could lead to a more permissive policy in the future depending on how chromium updates their policy for the stuff we don't specify above. https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/url_data_source.cc;l=112;drc=27d34700b83f381c62e3a348de2e6dfdc08364b8;bpv=1;bpt=1

we could just `return std::string();` here if we add the following to the above cases:

```
    case network::mojom::CSPDirectiveName::RequireTrustedTypesFor:
      return "require-trusted-types-for 'script';";
    case network::mojom::CSPDirectiveName::TrustedTypes:
      return "trusted-types;";
```

i think those are the only policies that are less permissive than what we have

Reply:
Fixed, cc @diracdeltas